### ISSUE TITLE
Simple gemm loop slower on 0.5

### ISSUE THREAD

**Issue by KristofferC** [JuliaLang/julia#15323]
If the stuff in here is just LLVM's "fault" and can't be fixed by us, feel free to close.

A typical triple nested loop between two square matrices is approx twice as slow on 0.5 than on 0.4. 

Script:
```jl
@noinline function matmul!{N}(C, A, B, ::Type{Val{N}})
    @inbounds for i in 1:N, j in 1:N, k in 1:N
        C[k, j] += A[k, i]  * B[i, j]
    end
end

function bench_simple_loop{N, T}(::Type{Val{N}}, ::Type{T})
    A = rand(T, N, N)
    C = similar(A)
    v = @elapsed for i in 1:10^5
        matmul!(C, A, A, Val{N})
    end
    return v
end
```

 `code_llvm` shows that 0.5 uses SIMD instructions even without the `@simd` macro., ~~0.4.3 does not use any SIMD instructions no matter what~~. 

Results for different versions with `Float32, Float64` using `@simd` or not on the inner loop (note `@simd` makes it significantly slower). Results are similar on https://github.com/JuliaLang/julia/pull/15139.

y-axis is time [s], x-axis is N for a [N x N] matrix
![matmul](https://cloud.githubusercontent.com/assets/1282691/13458170/4823f452-e06c-11e5-8996-e11efa8389e5.png)

**Comment by StefanKarpinski** [comment]
The y axis is seconds? Presumably not gigaflops since then there would be no issue..

**Comment by KristofferC** [comment]
Yes.. I think someone told me once that labeling axis is something an engineer should do. But yeah, it is in seconds.

**Comment by KristofferC** [comment]
Removing the `@inbounds` makes julia 0.5 faster. I believe the problem here is the SIMD instructions actually being slower than the scalar ones. 

With the inbounds julia 0.5 generates this for the loop

```
    vmovupd (%rdi), %xmm0
    vinsertf128 $1, 16(%rdi), %ymm0, %ymm0
    vmovupd (%rbx), %xmm1
    vinsertf128 $1, 16(%rbx), %ymm1, %ymm1
    vbroadcastsd    (%r8), %ymm2
    vmulpd  %ymm2, %ymm1, %ymm1
    vaddpd  %ymm1, %ymm0, %ymm0
    vextractf128    $1, %ymm0, 16(%rdi)
    vmovupd %xmm0, (%rdi)
```

I don't know enough assembly to say if this is good generated code but I CC some who might,  @ArchRobison @eschnett

**Comment by KristofferC** [comment]
Actually, I was wrong about julia 0.4 not using simd instructions. With the simd macro, julia 0.4 generates: 

```
    vbroadcastsd    (%r13,%rsi,8), %ymm0
    vmovupd (%rdi), %xmm1
    vinsertf128 $1, 16(%rdi), %ymm1, %ymm1
    vmulpd  %ymm0, %ymm1, %ymm0
    vmovupd (%rdx), %xmm1
    vinsertf128 $1, 16(%rdx), %ymm1, %ymm1
    vaddpd  %ymm0, %ymm1, %ymm0
    vextractf128    $1, %ymm0, %xmm1
    vmovupd %xmm1, 16(%rdx)
    vmovupd %xmm0, (%rdx)
```

and gets a speedup over not use `@simd`. Why would julia 0.5 get slowed down with `@simd` and julia 0.4 get sped up..

**Comment by eschnett** [comment]
Can you post the actual code used? The script listed above doesn't look right. The matrices are declared as types, not as arrays, `C` is not initialized, and `matmul!` is run only once for each set of argument types.

**Comment by ArchRobison** [comment]
What kind of difference are you seeing, as a time ratio?  What model processor?

The version of LLVM changed from 3.3 to 3.7.  Code generation that is hitting near several limits in the processor is tricky, and there invariably bets made that might not pay off in a particular circumstance, or the cost model in the compiler is missing some key item.  E.g., instruction decode can be a bottleneck, so modern Intel processors have a special small ]"uop" cache[(http://www.realworldtech.com/sandy-bridge/4/) that hold precoded instructions.

Also make sure that the matrices are always aligned the same.  Julia GC guarantees only 16-byte alignment, so ymm loads/stores may be misaligned arbitrarily.

**Comment by eschnett** [comment]
The disassembled code looks good for compiler-generated code. Missing optimization is a high-level loop transform that would transpose the input matrix `B`, or changing the loop order to make the `i`-loop the innermost loop. Given that Julia can't tell whether the arrays alias, LLVM is probably helpless here.

**Comment by KristofferC** [comment]
Running variations of this code: https://gist.github.com/KristofferC/4900df918446b07b7705

I get the following runtimes with 0.4 elapsed time as bsae

![figure_1](https://cloud.githubusercontent.com/assets/1282691/13647742/c055274c-e635-11e5-9565-5062827e50f3.png)

So without any macro decorator 0.5 is a constant factor faster. With `@inbounds` 0.5 is on average 2x slower, and with both `@inbounds` and `@simd` about 3x slower.

**Comment by KristofferC** [comment]
Oh and processor is  Intel(R) Core(TM) i5-2500K

**Comment by eschnett** [comment]
Is there a `@noalias` declaration available in LLVM that could be wrapped in Julia?

**Comment by KristofferC** [comment]
Looking at the generated code for `matmul!` in the above gist shows that 0.5 has a lot more code generated: https://gist.github.com/KristofferC/41006559c01e20c2f05e

**Comment by KristofferC** [comment]
There is also a 

```
    addq    $1, %r14
Source line: 4
L624:
    vmovsd  -8(%rdx,%r13,8), %xmm0  # xmm0 = mem[0],zero
    vmulsd  (%r8), %xmm0, %xmm0
    vaddsd  -8(%r9,%r13,8), %xmm0, %xmm0
    vmovsd  %xmm0, -8(%r9,%r13,8)
Source line: 3
    leaq    1(%r13), %r13
    cmpq    %r13, %r14
    jne L624
```

block in 0.5s code in addition to the vectorized code so is LLVM doing some run time checks first and decide whether to take the vectorized version or not? Maybe the vectorized code loop is not even running then?

**Comment by eschnett** [comment]
The vectorized code can only handle multiples of 4 (in your case). The scalar code cleans up the remaining iterations. In other words, if `N` is a multiple of 4, the code should run especially fast.

**Comment by KristofferC** [comment]
Ah, of course, thanks.

**Comment by ArchRobison** [comment]
FYI, LLVM does have [support for indicating non-aliasing](http://llvm.org/docs/LangRef.html#noalias-and-alias-scope-metadata).  From what I can tell, it's targeted at dealing with the level of detailed required to deal with inlining for Fortran and (C99 restrict).  It takes a fairly sophisticated user to get it right.  Compiler options that just say "nothing aliases in a function" scare me in a highly polymorphic language.

One thing the `@simd` macro is supposed to do is tell LLVM that it doesn't need a run-time alias check for vectorizing a  loop.  We need to look at the entire function body, not just the inner loop, to see what it's doing.

**Comment by ArchRobison** [comment]
Is the [gist](https://gist.github.com/KristofferC/41006559c01e20c2f05e#file-0-5) from code with or without `@simd`?

**Comment by carnaval** [comment]
@eschnett you should be able to make `@noalias` by generating fresh tbaa tags

**Comment by KristofferC** [comment]
@ArchRobison it is with `@simd`. Do you want the other versions as well?

**Comment by ArchRobison** [comment]
`@simd` version is enough for now.

**Comment by eschnett** [comment]
The `@simd` tag applies only to the innermost loop. If we want LLVM to be able to rearrange the other loops as well, then it needs to know more about non-aliasing. Maybe we should allow `@simd` also for loop nests? That corresponds to the `collapse` OpenMP directive.

**Comment by ArchRobison** [comment]
Yes, LLVM has the capability to [mark](http://llvm.org/docs/LangRef.html#llvm-mem-parallel-loop-access-metadata) multiple levels of loop as parallel.  So it seems reasonable to allow `@simd` on multiple levels.

**Comment by ArchRobison** [comment]
I suggest we look at the `N=1` case first and figure out why [the LLVM output](https://gist.github.com/ArchRobison/f913f6b6a9c5743ff17e) is so big.  I was expecting the value of N=1 to be propagated during compilation and the code to be munched down a straightline 1x1 matrix multiply.

**Comment by KristofferC** [comment]
That seems to happen on 0.4.3

**Comment by ArchRobison** [comment]
Something is going astray with `UnitRange`.  Here is a simple example of the root problem:
```
$ cat caboose.jl
@noinline function caboose{N}(::Type{Val{N}})
     last(1:N)
end

n = 42
caboose(Val{n})
@code_llvm caboose(Val{n})

$ julia caboose.jl

; Function Attrs: sspreq
define i64 @julia_caboose_23377(%jl_value_t*) #0 {
top:
  %1 = load i64, i64* inttoptr (i64 140644215072816 to i64*), align 16
  %2 = icmp sgt i64 %1, 0
  %3 = select i1 %2, i64 %1, i64 0
  ret i64 %3
}
```
With Julia 0.4 the code has just `ret i64 42`.  

@vtjnash: Is Julia 0.5 trying to do something clever with code generation for immutable objects that 0.4 didn't do?

**Comment by ArchRobison** [comment]
Never mind the question about immutable objects.  The issue is something about the way code is generated for parametric methods (perhaps to reduce code bloat?)  If I define `caboose` like this:
```
@noinline function caboose{N}(::Type{Val{N}})
     last(1:42)  # Like before, but with 1:42 instead of 1:N
end
```
then I get the trivial LLVM code that I expected.

**Comment by eschnett** [comment]
You could try a generated function instead; this ensures the function is always fully specialized:
```Julia
@generated function caboose{N}(::Type{Val{N}})
    quote
        $(Expr(:meta, :noinline))
         last(1:N)
    end
end
```

**Comment by vtjnash** [comment]
that codegen regression will be fixed by https://github.com/JuliaLang/julia/pull/15344 (i forgot that wasn't merged yet).

but please don't use `Val` or `@generated` to write a function like this. the constant inlined version will generate very bad dispatch and codegen over-specialization issues.

**Comment by KristofferC** [comment]
For reference here is the same analysis as in https://github.com/JuliaLang/julia/issues/15323#issuecomment-194468442 on #15344:

![figure_1](https://cloud.githubusercontent.com/assets/1282691/13667242/beff6fa8-e6b5-11e5-93e0-a7d954002da0.png)

**Comment by ArchRobison** [comment]
It appears that without hacking on LLVM, `@simd` is not going to help much with short fixed-trip-count loops.  Root problem is the typical "phase order" problem in compilers: the pass order is based on bets that are sometimes lost.

Details: With #15344, LLVM has precise trip-count information.  Unfortunately it uses that information to completely unroll the inner loop for small N, and unfortunately in a way that loses information that SLP vectorization needs to vectorize the resulting straight-line code.  We could mark `@simd` loops with LLVM's metadata that says to not unroll the loop, but I suspect that's deleterious to the vectorizer's interleaving optimization, which is critical for large loops.

We could of course add the metadata and strip it just before the vectorizer runs, but adding another pass seems like too much overhead given Julia's current compile-time blues.  Maybe patching the LLVM unroller to avoid monkeying with parallel loops would help, but then we're into betting territory.

Given this state of LLVM, `SIMD.jl` seems like a better approach for implementing small fixed-sized matrix multiplications.

**Comment by Keno** [comment]
> unfortunately in a way that loses information that SLP vectorization needs to vectorize the resulting straight-line code

can this be fixed in LLVM? That sounds like a bug.

**Comment by ArchRobison** [comment]
The problem seems to be inherent in the LLVM representation.  This [old posting](https://groups.google.com/forum/#!topic/llvm-dev/n-ia9kJyg6w) touches on the subject, and proposed an extension of the metadata.  

I'm not entirely sure that disabling loop unrolling will interfere with the vectorizer interleaving the iterations.  The logic in the vectorizer is hard to follow on this point.

**Comment by KristofferC** [comment]
As https://github.com/JuliaLang/julia/issues/15323#issuecomment-194787201 shows the situation is now much better than in the first post. The issue regarding the vectorizer failing when the loop is unrolled could maybe deserve a separate issue.
